File: nodejs-tools/nodejs-compiler-tools/formatter-xml/src/format_xml_element_attribute.test.ts
import { expect, test } from "vitest";
import { Token } from "@vighnesh153/lexer-core";
import { XmlTokenType } from "@vighnesh153/lexer-xml";
import { formatXmlElementAttribute } from "./format_xml_element_attribute.ts";

function createToken(
  tokenLiteral: string,
  tokenType: XmlTokenType,
): Token<XmlTokenType> {
  return {
    lineNumber: 1,
    columnNumber: 1,
    tokenLiteral,
    tokenType,
  } satisfies Token<XmlTokenType>;
}

test("should format attribute with single namespace", () => {
  expect(
    formatXmlElementAttribute({
      namespaces: [createToken("pokemon", XmlTokenType.Identifier)],
      value: createToken("pikachu", XmlTokenType.StringLiteral),
    }),
  ).toMatchInlineSnapshot(`"pokemon="pikachu""`);
});

test("should format attribute with multiple namespaces", () => {
  expect(
    formatXmlElementAttribute({
      namespaces: [
        createToken("pokemon", XmlTokenType.Identifier),
        createToken("electric", XmlTokenType.Identifier),
        createToken("yellow", XmlTokenType.Identifier),
      ],
      value: createToken("pikachu", XmlTokenType.StringLiteral),
    }),
  ).toMatchInlineSnapshot(`"pokemon:electric:yellow="pikachu""`);
});

File: tools-jvm/XmlFormatter/src/test/kotlin/dev/vighnesh153/xml/formatter/fmt/FormatXmlElementAttributeTest.kt
package dev.vighnesh153.xml.formatter.fmt

import dev.vighnesh153.xml.formatter.ast.XmlElementAttribute
import dev.vighnesh153.xml.formatter.lexer.Token
import dev.vighnesh153.xml.formatter.lexer.TokenType
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class FormatXmlElementAttributeTest {
    @Test
    fun formatsWithSingleNamespace() {
        val xmlElementAttribute = XmlElementAttribute(
            namespaces = listOf(
                createToken(TokenType.Identifier, "pokemon"),
            ),
            value = createToken(TokenType.StringLiteral, "pikachu")
        )

        assertEquals(
            """
                pokemon="pikachu"
            """.trim(),
            xmlElementAttribute.format()
        )
    }

    @Test
    fun formatsWithMultipleNamespaces() {
        val xmlElementAttribute = XmlElementAttribute(
            namespaces = listOf(
                createToken(TokenType.Identifier, "pokemon"),
                createToken(TokenType.Identifier, "electric"),
                createToken(TokenType.Identifier, "yellow"),
            ),
            value = createToken(TokenType.StringLiteral, "pikachu")
        )

        assertEquals(
            """
                pokemon:electric:yellow="pikachu"
            """.trim(),
            xmlElementAttribute.format()
        )
    }

    private fun createToken(tokenType: TokenType, tokenLiteral: String) = Token(
        tokenType = tokenType,
        tokenLiteral = tokenLiteral,
        lineNumber = 1,
        columnNumber = 1,
    )
}